uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-k8s/src/main/java/org/uberfire/java/nio/fs/k8s/K8SWatchService.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.uberfire.java.nio.fs.k8s;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.uberfire.java.nio.file.ClosedWatchServiceException;
+import org.uberfire.java.nio.file.InterruptedException;
+import org.uberfire.java.nio.file.Path;
+import org.uberfire.java.nio.file.WatchKey;
+import org.uberfire.java.nio.file.WatchService;
+import org.uberfire.java.nio.fs.cloud.CloudClientFactory;
+
+import static org.uberfire.java.nio.fs.k8s.K8SFileSystemConstants.CFG_MAP_LABEL_FSOBJ_APP_KEY;
+import static org.uberfire.java.nio.fs.k8s.K8SFileSystemUtils.APP_NAME;
+import static org.uberfire.java.nio.fs.k8s.K8SFileSystemUtils.getPathByFsObjCM;
+import static org.uberfire.java.nio.fs.k8s.K8SFileSystemUtils.mapActionToKind;
+
+public class K8SWatchService implements WatchService {
+
+    private static final Logger logger = LoggerFactory.getLogger(K8SWatchService.class);
+    private final CloudClientFactory ccf;
+    private final K8SFileSystem fs;
+    private final BlockingQueue<WatchKey> buckets = new LinkedBlockingQueue<>();
+    private final Map<Path, WatchKey> registrations = new ConcurrentHashMap<>();
+
+    private final CompletableFuture<Void> closed = new CompletableFuture<>();
+
+    public K8SWatchService(K8SFileSystem fs) {
+        this.fs = fs;
+        this.ccf = (CloudClientFactory) fs.provider();
+        Executors.newSingleThreadExecutor().execute(() -> 
```

Comment:
Please use DescriptiveRunnable class here, because if we have problems with threads it is impossible to follow them up if they don't have a proper name

Reply:
@adrielparedes That's a good point! 
Since we have to be thoughtful on the dependencies just in case when need to refactor k8sfs into a standalone utility, how about this alternative solution to name the thread by leveraging [guava](https://guava.dev/releases/snapshot/api/docs/com/google/common/util/concurrent/ThreadFactoryBuilder.html#setNameFormat-java.lang.String-).

` Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
      .setNameFormat("k8sfs-cm-watcher-thread-%d").build()).execute(() -> 
`
If it is ok, I'll update the PR accordingly.

Reply:
+1

Reply:
Watcher thread being named, see the following log captured from integration test run.
```
09:33:01.081 [k8sfs-cm-watcher-thread-0] INFO  o.u.java.nio.fs.k8s.K8SWatchService - K8SFileSystem WatchService is starting to watch K8SFileSystem ConfigMap in namespace: [k8sfsp-test]
09:33:01.093 [k8sfs-cm-watcher-thread-0] INFO  o.u.java.nio.fs.k8s.K8SWatchService - K8SFileSystem ConfigMap Watcher thread started.
09:33:01.143 [main] INFO  o.u.java.nio.fs.k8s.K8SWatchService - K8SFileSystem WatchService is closing.
09:33:01.143 [main] INFO  o.u.java.nio.fs.k8s.K8SWatchService - K8SFileSystem WatchService closed normally.
09:33:01.143 [k8sfs-cm-watcher-thread-0] INFO  o.u.java.nio.fs.k8s.K8SWatchService - K8SFileSystem ConfigMap Watcher thread terminated.
09:33:01.144 [k8sfs-cm-watcher-thread-0] INFO  o.u.java.nio.fs.k8s.K8SWatchService - K8SFileSystem ConfigMap Watcher closed.
```